### PR TITLE
Modify the mission order: upload, arm, start

### PR DESCRIPTION
**Pull request by JonasVautherin**
Quickly trying some changes in the mission start, and noting a few points:

* The mission needs to be uploaded to the drone before it can be started, so the order is:
    1. upload
    2. arm
    3. start mission
* When the drone is armed, if nothing happens for a few seconds, it automatically disarms
* On the cloud simulator, uploading one mission item can take 200-400ms, meaning that uploading hundreds of items may take a little while. That's a limitation of the cloud sim, on a real drone it is much faster. This means that if you arm, and then upload the mission, it will automatically disarm while you upload, that's expected.
* No need to takeoff before starting a mission, all you need is that the drone is armed

If you want to play with a small java example uploading many waypoints, we recently discussed that [here](https://github.com/mavlink/MAVSDK-Java/issues/67#issuecomment-843877587). Also the order is shown in the [RunMission.java](https://github.com/mavlink/MAVSDK-Java/blob/main/examples/java-client/src/main/java/io/mavsdk/example/RunMission.java#L28-L32) example :+1:.

I did not check if the tests were still passing, the intent here is just to show the best order for starting a mission :blush:.

Hope it helps!

### CODE REVIEW

**Comment by GabrielFleischer** [comment]
I adapted the change with our app, but it seems that I cannot upload a mission, I have timeout issues.

> 2021-05-23 17:49:26.753 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 527 not matching our current command: 525
> 2021-05-23 17:49:26.760 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.507153 s, retries to do: 3  (525).
> 2021-05-23 17:49:27.441 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 525 not matching our current command: 2504
> 2021-05-23 17:49:27.540 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.505023 s, retries to do: 3  (2504).
> 2021-05-23 17:49:28.079 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 2504 not matching our current command: 2505
> 2021-05-23 17:49:28.128 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.504555 s, retries to do: 3  (2505).
> 2021-05-23 17:49:28.409 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 2505 not matching our current command: 527
> 2021-05-23 17:49:31.939 25728-25846/ch.epfl.sdp.drone3d D/Mavsdk: MAVLink: critical: Operation timeout
> 2021-05-23 17:49:49.133 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.509758 s, retries to do: 3  (525).
> 2021-05-23 17:49:49.662 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 525 not matching our current command: 2504
> 2021-05-23 17:49:49.755 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.50209 s, retries to do: 3  (2504).
> 2021-05-23 17:49:50.250 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 2504 not matching our current command: 2505
> 2021-05-23 17:49:50.333 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.509232 s, retries to do: 3  (2505).
> 2021-05-23 17:49:50.788 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 2505 not matching our current command: 527
> 2021-05-23 17:49:51.020 25728-25839/ch.epfl.sdp.drone3d W/Mavsdk: sending again after 0.501568 s, retries to do: 3  (527).
> 2021-05-23 17:49:51.390 25728-25846/ch.epfl.sdp.drone3d W/Mavsdk: Command ack 527 not matching our current command: 525

I tried with big and small missions, and the only difference is that the timeout happens sooner on small missions.
So I don't think this is actually related to the items taking too long.

I will try to play with it more to see if I can find the issue.

**Comment by GabrielFleischer** [comment]
I managed to pass the upload mission by recreating a simulation.

But it got disarmed by the auto-preflight system again.

**Comment by JonasVautherin** [comment]
Hmm it was working for me :thinking:. Are you running from a phone or from the Android simulator? Could be that the Android simulator adds delay on top of it, making the connection even more brittle (which would be something to improve in MAVSDK, e.g. by tuning the timeout delay in those conditions :thinking:).

Also you can try to run a simulator locally (see https://github.com/jonasvautherin/px4-gazebo-headless), though this one does not have the camera controls (it exposes a video stream, but it won't take pictures). Still interesting to check the mission stuff.

**Comment by GabrielFleischer** [comment]
Using a local simulator worked like a charm. So I am confident this was related to the simulator.

**Review by GabrielFleischer — APPROVED** [review]

**Review by GabrielFleischer — APPROVED** [review]

File: app/src/main/java/ch/epfl/sdp/drone3d/service/impl/drone/DroneExecutorImpl.kt [hunk 1]
```diff
@@ -220,6 +175,25 @@ class DroneExecutorImpl(
...
+                    data.getMutableDroneStatus().postValue(SENDING_ORDER)
+                }
+                .andThen(droneInstance.mission.uploadMission(missionPlan)
+                        .andThen(droneInstance.mission.startMission())
+                        .doOnComplete {
+                            data.getMutableDroneStatus().postValue(EXECUTING_MISSION)
+                            data.getMutableMission().postValue(null)
+                            data.getMutableMissionPaused().postValue(false)
+                            ToastHandler.showToastAsync(ctx, msg)
+                        })
```
Comment (Jorque) [inline app/src/main/java/ch/epfl/sdp/drone3d/service/impl/drone/DroneExecutorImpl.kt:195]: There are extra unnecessary spaces here
```suggestion
                .andThen(droneInstance.mission.startMission())
                .doOnComplete {
                    data.getMutableDroneStatus().postValue(EXECUTING_MISSION)
                    data.getMutableMission().postValue(null)
                    data.getMutableMissionPaused().postValue(false)
                    ToastHandler.showToastAsync(ctx, msg)
                })
```

**Review by Jorque — CHANGES_REQUESTED** [review]
Nice job! Just remove these extra spaces and it's good

**Review by Jorque — APPROVED** [review]
I guess I could just do it myself

**Pull request merged**